### PR TITLE
meson.eclass: wire up LTO support directly into the meson options

### PR DESCRIPTION
**Pull request by eli-schwartz**
meson's builtin LTO support allows meson to introspect whether LTO is enabled and do some fancy things, such as forcing LTO off for a single target that is known to be special(ly bad) and not support LTO.

Includes:
- meson.eclass: refactor src_configure into a setter function
- demonstration of how this works for distutils-r1

/cc @floppym

### CODE REVIEW

**Comment by eli-schwartz** [comment]
Thanks. Please feel free to merge the first three commits and mark it as closing this PR, as I plan to handle the distutils-r1 stuff separately anyway and only included it here to help provide context on how the meson.eclass changes could be used.

It's either that or when I get home I'll force push this PR to remove the distutils-r1 commits anyway :D

File: eclass/meson.eclass [hunk 1]
```diff
@@ -277,15 +277,12 @@ meson_feature() {
...
+# @FUNCTION: setup_meson_src_configure
 # @DESCRIPTION:
-# This is the meson_src_configure function.
-meson_src_configure() {
-	debug-print-function ${FUNCNAME} "$@"
-
-	[[ -n "${NINJA_DEPEND}" ]] || ewarn "Unknown value '${NINJA}' for \${NINJA}"
-
+# Calculate the command line which meson should use, and other relevant
+# variables. Invoke via "${mesonargs[@]}" in the calling environment.
```
Comment (mgorny) [inline eclass/meson.eclass:283]: I don't exactly like the idea of "set `mesonargs` and possibly more variables that will change in the future".

File: eclass/meson.eclass [hunk 2]
```diff
@@ -379,13 +369,32 @@ meson_src_configure() {
 	python_export_utf8_locale
```
Comment (mgorny) [inline eclass/meson.eclass:397]: This one's redundant to what python eclasses do.

File: eclass/meson.eclass [hunk 3]
```diff
@@ -283,6 +283,36 @@ meson_feature() {
 # variables. Invoke via "${mesonargs[@]}" in the calling environment.
 # This function is called from meson_src_configure.
 setup_meson_src_configure() {
+	local ltoflags=()
```
Comment (mgorny) [inline eclass/meson.eclass:286]: Is there a reason to introduce a separate `ltoflags` for this and not simply put it straight to `mesonargs`? I don't think the option order really matters here.

File: eclass/distutils-r1.eclass [hunk 4]
```diff
@@ -197,6 +197,10 @@ _DISTUTILS_R1_ECLASS=1
 inherit flag-o-matic
 inherit multibuild multilib multiprocessing ninja-utils toolchain-funcs
 
+if [[ ${DISTUTILS_USE_PEP517} = meson-python ]]; then
```
Comment (mgorny) [inline eclass/distutils-r1.eclass:200]: Already told you to use `==` for consistency.

File: eclass/distutils-r1.eclass [hunk 5]
```diff
@@ -1386,9 +1390,11 @@ distutils_pep517_install() {
 			)
 			;;
 		meson-python)
+			local mesonargs=()
```
Comment (mgorny) [inline eclass/distutils-r1.eclass:1393]: Also add `local` for all the other variables set by `setup_meson_src_configure`.

File: eclass/meson.eclass [hunk 6]
```diff
@@ -314,8 +311,7 @@ meson_src_configure() {
 		: "${BUILD_PKG_CONFIG_PATH:=${PKG_CONFIG_PATH}}"
 	fi
 
-	local mesonargs=(
-		meson setup
+	mesonargs=(
```
Comment (mgorny) [inline eclass/meson.eclass:344]: This is no longer a local variable, so technically speaking it would be more correct to make it uppercase.

File: eclass/distutils-r1.eclass [hunk 7]
```diff
@@ -1815,16 +1815,15 @@ distutils-r1_run_phase() {
 	fi
 
 	# Set up build environment, bug #513664.
-	local -x AR=${AR} CC=${CC} CPP=${CPP} CXX=${CXX}
```
Comment (mgorny) [inline eclass/distutils-r1.eclass:1812]: I already NAK-d this, and the whole underlying code is changed in #35554.

**Review by mgorny — CHANGES_REQUESTED** [review]

File: eclass/distutils-r1.eclass (hunk 7 above)
Comment (eli-schwartz) [reply eclass/distutils-r1.eclass:1812]: I'm not certain why you're pointing this out since I already clarified I don't intend to have distutils-r1 changes merged yet, *and* this PR predates yours *and* I reviewed your PR and generally agree with your approach.

Even though your approach was only motivated by my proposal. :)

tl;dr I already agree with you. No worries.

File: eclass/meson.eclass (hunk 2 above)
Comment (eli-schwartz) [reply eclass/meson.eclass:397]: I didn't specifically change this but I'd be happy to remove it as I wasn't sure myself why it should be needed.

File: eclass/meson.eclass (hunk 3 above)
Comment (eli-schwartz) [reply eclass/meson.eclass:286]: I am not sure why I did that but I think it may be a side effect of having refactored code around without thinking too deeply about whether the order matters.

Let me take another look.

File: eclass/meson.eclass (hunk 6 above)
Comment (eli-schwartz) [reply eclass/meson.eclass:344]: Sounds reasonable enough.

I'm very much not attached to any particular name here.

File: eclass/meson.eclass [hunk 8]
```diff
@@ -379,13 +369,32 @@ meson_src_configure() {
...
+# @USAGE: [extra meson arguments]
+# @DESCRIPTION:
+# This is the meson_src_configure function.
+meson_src_configure() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	[[ -n "${NINJA_DEPEND}" ]] || ewarn "Unknown value '${NINJA}' for \${NINJA}"
 
 	(
+		setup_meson_src_configure "$@"
```
Comment (swegener) [inline eclass/meson.eclass:416]: Calling `setup_meson_src_configure` inside the `() || die`, the variables are not exported for re-use in other phases, especially `BUILD_DIR`, which is required in `meson_install`.

Comment (swegener) [reply eclass/meson.eclass:416]: Thanks, see f2e1792ded67e62edc85e92e84f65d773a8f05bd
